sdk/python/symbolchain/nem/KeyPair.py

```diff
@@ -34,10 +84,48 @@ def __init__(self, public_key):
 		"""Creates a verifier from a public key."""
 		self._pk = public_key.bytes
 
+	@staticmethod
+	def is_canonical_s(encoded_s):
+		# reject zero S
+		if encoded_s == bytes(32):
+			return False
+
+		# reject non-reduced S
+		reduced = crypto_core_ed25519_scalar_reduce(encoded_s + bytes(32))
+		return reduced == encoded_s
+
 	def verify(self, message, signature):
 		"""Verifies a message signature."""
-		try:
-			ed25519.checkvalid_hash(signature.bytes, message, self._pk, sha3.keccak_512)
-			return True
-		except (ValueError, ed25519.SignatureMismatch):
+
+		encoded_r = signature.bytes[:32]
+		encoded_s = signature.bytes[32:]
+
+		if not self.is_canonical_s(encoded_s):
+			return False
+
+		if bytes(32) == self._pk:
 			return False
+
+		# h = H(encodedR || public || data)
+		hashobj = sha3.keccak_512()
+		hashobj.update(encoded_r)
+		hashobj.update(self._pk)
+		hashobj.update(message)
+		hram = hashobj.digest()
+		hram = crypto_core_ed25519_scalar_reduce(hram)
+
+		# note: this check covers following:
+		#  * (y) coord is canonical (< 2^255 - 19)
+		#  * has small order - this is slightly stronger check,
+		#    we don't do that in the client as we verify sigs with pub keys that are known to belong to accounts
```

Comment:
could client generate small order key ?

Reply:
(discussed on discord)
tl;dr, should not be an issue on server part and subgroup check should be covering this as well